### PR TITLE
minisat: use size_t to index clauses

### PR DESCRIPTION
**Pull request by tautschnig**
This enables Minisat to use more than 16GB of memory for storing clauses.

### CODE REVIEW

**Comment by smowton** [comment]
Please add this to CMake as well -- I believe you'll want to use `ExternalProject_Add_Step` in https://github.com/diffblue/cbmc/blob/develop/src/solvers/CMakeLists.txt, probably with arguments `DEPENDEES patch DEPENDERS build` to get the step sequence download -> patch -> patch64 -> build.

**Comment by smowton** [comment]
(Docs: https://cmake.org/cmake/help/v3.8/module/ExternalProject.html)

**Comment by tautschnig** [comment]
@kroening Any guidance on why CodeBuild is unhappy here?

**Review by smowton — APPROVED** [review]
Checked cmake step; seems to be working as we'd hope

**Comment by kroening** [comment]
This fails the "lots_of_local_variables" test. My suspicion is that some ordering here is undefined.

**Comment by kroening** [comment]
BTW, does this impact performance on smaller instances?

File: scripts/minisat64.patch [hunk 1]
```diff
@@ -0,0 +1,1770 @@
...
+ 
+             if (!seen[var(q)] && level(var(q)) > 0){
+@@ -301,8 +301,8 @@
+         }
+         
+         // Select next clause to look at:
+-        while (!seen[var(trail[index--])]);
+-        p     = trail[index+1];
++        while (!seen[var(trail[--index])]);
++        p     = trail[index];
```
Comment (martin-cs) [inline scripts/minisat64.patch:118]: Ummm... why change this?

File: scripts/minisat64.patch [hunk 2]
```diff
@@ -0,0 +1,1770 @@
...
++public:
++    typedef uint32_t AbsLevel;
++private:
++#if SIZE_MAX == UINT32_MAX
++    union { Lit lit; float act; AbsLevel abs; CRef rel; } data[0];
++#else
++    static_assert(sizeof(Lit) == sizeof(float), "float has the size of Lit");
++    static_assert(sizeof(Lit) == sizeof(AbsLevel), "AbsLevel has the size of Lit");
++    static_assert(sizeof(Lit) == sizeof(uint32_t), "Lit is uint32_t");
++    static_assert(2 * sizeof(Lit) == sizeof(CRef), "CRef is 2*Lit");
```
Comment (martin-cs) [inline scripts/minisat64.patch:629]: This concerns me; do we have any performance data on the effects of this?

**Review by martin-cs — CHANGES_REQUESTED** [review]
I'd like to see some performance numbers before we commit to this.

**Comment by NlightNFotis** [comment]
Hi @tautschnig, this looks useful! Could we have it rebased on top of `develop` please?

**Comment by thomasspriggs** [comment]
This may be something of a tangent to the subject of the review of this PR. However I think it is worth considering whether a set of patches like this is a good way of maintaining our variation of MiniSat; given that git a commit essentially is a patch. So if we could keep the patches as commits we would have better/standardised tools for maintaining version history and resolving any merge conflicts which occur.

**Comment by martin-cs** [comment]
@thomasspriggs having Minisat as a git submodule with patches would also remove the download-during-build gotcha.

**Comment by tautschnig** [comment]
Closing as it would not be right for us to carry such a patch. Could be upstreamed to [conp/merg](https://github.com/conp-solutions/mergesat) instead.

**Pull request closed without merging**